internal/debug/debug.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import (
+	"fmt"
+	"html/template"
+	"net/http"
+	"path"
+
+	"go.uber.org/yarpc/internal/introspection"
+)
+
+type httpMux interface {
+	Handle(pattern string, handler http.Handler)
+	HandleFunc(pattern string, handler func(http.ResponseWriter, *http.Request))
+}
+
+// IntrospectionProvider expects a set of methods to retrieve YARPC specific
+// introspection informations.
+type IntrospectionProvider interface {
+	Dispatchers() []introspection.DispatcherStatus
+	DispatchersByName(name string) []introspection.DispatcherStatus
+	PackageVersions() []introspection.PackageVersion
+}
+
+type page struct {
+	path    string
+	handler func(w http.ResponseWriter, req *http.Request,
+		insp IntrospectionProvider) interface{}
+	html string
+	tmpl *template.Template
+}
+
+// Pages holds a bag of yarpc debug pages. Binding a set of pages to an
+// introspection provider.
+type Pages struct {
+	insp  IntrospectionProvider
+	pages map[string]*page
+}
+
+func (dp *Pages) registerPage(page *page) {
+	funcmap := map[string]interface{}{
+		"pathBase":    path.Base,
+		"wrapIDLtree": wrapIDLTree,
+	}
+
+	// We do not .Clone() the base template, but reparse it every time. Because
+	// of a race condition/memory leak when template.Clone() is used in
+	// conjunction with template blocks on Go<=1.8.
+	base := template.Must(template.New("base").Funcs(funcmap).Parse(baseHTML))
```

Comment:
it must never ever fail. a failure here is a programming error, and will be caught when running tests.

I am not using globals mostly for code organisation and working around the race condition/leak bug.